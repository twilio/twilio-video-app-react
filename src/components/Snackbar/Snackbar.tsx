import React from 'react';
import clsx from 'clsx';
import CloseIcon from '@material-ui/icons/Close';
import ErrorIcon from '../../icons/ErrorIcon';
import { IconButton, makeStyles, Theme, Typography } from '@material-ui/core';
import MUISnackbar from '@material-ui/core/Snackbar';
import WarningIcon from '../../icons/WarningIcon';
import InfoIcon from '../../icons/InfoIcon';

interface SnackbarProps {
  headline: string;
  message: string | React.ReactNode;
  variant?: 'error' | 'warning' | 'info';
  open: boolean;
  handleClose?: () => void;
}

const useStyles = makeStyles((theme: Theme) => ({
  container: {
    display: 'flex',
    justifyContent: 'space-between',
    width: '400px',
    minHeight: '50px',
    background: 'white',
    padding: '1em',
    borderRadius: '3px',
    boxShadow: '0 12px 24px 4px rgba(40,42,43,0.2)',
    [theme.breakpoints.down('sm')]: {
      width: '100%',
    },
  },
  contentContainer: {
    display: 'flex',
    lineHeight: 1.8,
  },
  iconContainer: {
    display: 'flex',
    padding: '0 1.3em 0 0.3em',
    transform: 'translateY(3px)',
  },
  headline: {
    fontWeight: 'bold',
  },
  error: {
    borderLeft: '4px solid #D61F1F',
  },
  warning: {
    borderLeft: '4px solid #E46216',
  },
  info: {
    borderLeft: '4px solid #0263e0',
  },
}));

export default function Snackbar({ headline, message, variant, open, handleClose }: SnackbarProps) {
  const classes = useStyles();

  const handleOnClose = (_: React.SyntheticEvent | React.MouseEvent, reason?: string) => {
    if (reason === 'clickaway') {
      return;
    }

    handleClose?.();
  };

  return (
    <MUISnackbar
      anchorOrigin={{
        vertical: 'top',
        horizontal: 'right',
      }}
      open={open}
      onClose={handleOnClose}
      autoHideDuration={10000}
    >
      <div
        className={clsx(classes.container, {
          [classes.error]: variant === 'error',
          [classes.warning]: variant === 'warning',
          [classes.info]: variant === 'info',
        })}
      >
        <div className={classes.contentContainer}>
          <div className={classes.iconContainer}>
            {variant === 'warning' && <WarningIcon />}
            {variant === 'error' && <ErrorIcon />}
            {variant === 'info' && <InfoIcon />}
          </div>
          <div>
            <Typography variant="body1" className={classes.headline} component="span">
              {headline}
            </Typography>
            <Typography variant="body1" component="span">
              {' '}
              {message}
            </Typography>
          </div>
        </div>
        <div>
          {handleClose && (
            <IconButton size="small" aria-label="close" color="inherit" onClick={handleClose}>
              <CloseIcon fontSize="small" />
            </IconButton>
          )}
        </div>
      </div>
    </MUISnackbar>
  );
}
