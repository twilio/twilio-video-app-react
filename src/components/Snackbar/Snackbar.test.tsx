import { shallow } from 'enzyme';
import React from 'react';
import Snackbar from './Snackbar';

describe('the Snackbar component', () => {
  it('should render correctly with "warning" variant', () => {
    const wrapper = shallow(
      <Snackbar variant="warning" headline="Test Headline" message="Test Message" handleClose={() => {}} open={true} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with "error" variant', () => {
    const wrapper = shallow(
      <Snackbar variant="error" headline="Test Headline" message="Test Message" handleClose={() => {}} open={true} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with "info" variant', () => {
    const wrapper = shallow(
      <Snackbar variant="info" headline="Test Headline" message="Test Message" handleClose={() => {}} open={true} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with no handleClose function provided', () => {
    const wrapper = shallow(<Snackbar variant="error" headline="Test Headline" message="Test Message" open={true} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('the handleClose function', () => {
    beforeEach(jest.clearAllMocks);

    const mockHandleClose = jest.fn();
    const wrapper = shallow(
      <Snackbar
        variant="warning"
        headline="Test Headline"
        message="Test Message"
        handleClose={mockHandleClose}
        open={true}
      />
    );

    it('should be called when the onClose function is called', () => {
      wrapper.prop('onClose')();
      expect(mockHandleClose).toHaveBeenCalled();
    });

    it('should be called when the onClose function is called with the "clickaway" reason', () => {
      wrapper.prop('onClose')({}, 'clickaway');
      expect(mockHandleClose).not.toHaveBeenCalled();
    });
  });
});
