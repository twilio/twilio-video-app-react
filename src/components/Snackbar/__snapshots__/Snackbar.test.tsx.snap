// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Snackbar component should render correctly with "error" variant 1`] = `
<WithStyles(ForwardRef(Snackbar))
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  autoHideDuration={10000}
  onClose={[Function]}
  open={true}
>
  <div
    className="makeStyles-container-1 makeStyles-error-5"
  >
    <div
      className="makeStyles-contentContainer-2"
    >
      <div
        className="makeStyles-iconContainer-3"
      >
        <ErrorIcon />
      </div>
      <div>
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-headline-4"
          component="span"
          variant="body1"
        >
          Test Headline
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          component="span"
          variant="body1"
        >
           
          Test Message
        </WithStyles(ForwardRef(Typography))>
      </div>
    </div>
    <div>
      <WithStyles(ForwardRef(IconButton))
        aria-label="close"
        color="inherit"
        onClick={[Function]}
        size="small"
      >
        <Memo(ForwardRef(CloseIcon))
          fontSize="small"
        />
      </WithStyles(ForwardRef(IconButton))>
    </div>
  </div>
</WithStyles(ForwardRef(Snackbar))>
`;

exports[`the Snackbar component should render correctly with "info" variant 1`] = `
<WithStyles(ForwardRef(Snackbar))
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  autoHideDuration={10000}
  onClose={[Function]}
  open={true}
>
  <div
    className="makeStyles-container-1 makeStyles-info-7"
  >
    <div
      className="makeStyles-contentContainer-2"
    >
      <div
        className="makeStyles-iconContainer-3"
      >
        <InfoIcon />
      </div>
      <div>
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-headline-4"
          component="span"
          variant="body1"
        >
          Test Headline
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          component="span"
          variant="body1"
        >
           
          Test Message
        </WithStyles(ForwardRef(Typography))>
      </div>
    </div>
    <div>
      <WithStyles(ForwardRef(IconButton))
        aria-label="close"
        color="inherit"
        onClick={[Function]}
        size="small"
      >
        <Memo(ForwardRef(CloseIcon))
          fontSize="small"
        />
      </WithStyles(ForwardRef(IconButton))>
    </div>
  </div>
</WithStyles(ForwardRef(Snackbar))>
`;

exports[`the Snackbar component should render correctly with "warning" variant 1`] = `
<WithStyles(ForwardRef(Snackbar))
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  autoHideDuration={10000}
  onClose={[Function]}
  open={true}
>
  <div
    className="makeStyles-container-1 makeStyles-warning-6"
  >
    <div
      className="makeStyles-contentContainer-2"
    >
      <div
        className="makeStyles-iconContainer-3"
      >
        <WarningIcon />
      </div>
      <div>
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-headline-4"
          component="span"
          variant="body1"
        >
          Test Headline
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          component="span"
          variant="body1"
        >
           
          Test Message
        </WithStyles(ForwardRef(Typography))>
      </div>
    </div>
    <div>
      <WithStyles(ForwardRef(IconButton))
        aria-label="close"
        color="inherit"
        onClick={[Function]}
        size="small"
      >
        <Memo(ForwardRef(CloseIcon))
          fontSize="small"
        />
      </WithStyles(ForwardRef(IconButton))>
    </div>
  </div>
</WithStyles(ForwardRef(Snackbar))>
`;

exports[`the Snackbar component should render correctly with no handleClose function provided 1`] = `
<WithStyles(ForwardRef(Snackbar))
  anchorOrigin={
    Object {
      "horizontal": "right",
      "vertical": "top",
    }
  }
  autoHideDuration={10000}
  onClose={[Function]}
  open={true}
>
  <div
    className="makeStyles-container-1 makeStyles-error-5"
  >
    <div
      className="makeStyles-contentContainer-2"
    >
      <div
        className="makeStyles-iconContainer-3"
      >
        <ErrorIcon />
      </div>
      <div>
        <WithStyles(ForwardRef(Typography))
          className="makeStyles-headline-4"
          component="span"
          variant="body1"
        >
          Test Headline
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Typography))
          component="span"
          variant="body1"
        >
           
          Test Message
        </WithStyles(ForwardRef(Typography))>
      </div>
    </div>
    <div />
  </div>
</WithStyles(ForwardRef(Snackbar))>
`;
